pkg/utils/Idpool.go

```diff
@@ -0,0 +1,154 @@
+	id._idsInUse = make(map[interface{}]uint32)
+	id._idsForReuse = make(map[interface{}]uint32)
+	id._refs = make(map[uint32][]interface{})
+	return id
+}
+
+func (ip *IDPool) _assignID(key interface{}) uint32 {
+	// Check if there was an id assigned for that key earlier
+	id := ip._idsForReuse[key]
+	if !reflect.ValueOf(id).IsZero() {
+		// Re-use the old id
+		delete(ip._idsForReuse, key)
+	} else {
+		if len(ip._unusedIDs) != 0 {
+			// Pick an unused id
+			id = ip._unusedIDs[0]
+			ip._unusedIDs = append(ip._unusedIDs[:0], ip._unusedIDs[1:]...)
+		} else {
+			if len(ip._idsForReuse) != 0 {
+				// Pick one of the ids earlier used for another key
+				for oldKey := range ip._idsForReuse {
+					delete(ip._idsForReuse, oldKey)
+					break
+				}
+			} else {
+				// No id left
+				log.Printf("IDPool: Failed to allocate id for %+v. No free ids in pool.", key)
+				return 0
+			}
+		}
+	}
+	// Store the assigned id, if any
+	if !reflect.ValueOf(id).IsZero() {
+		ip._idsInUse[key] = id
+	}
+	return id
+}
+
+// GetID get the mod ptr id from pool
+func (ip *IDPool) GetID(key interface{}, ref interface{}) (uint32, uint32) {
+	id := ip._idsInUse[key]
+	if reflect.ValueOf(id).IsZero() {
+		// Assign a free id for the key
+		id = ip._assignID(key)
+		if id == 0 {
+			return 0, 0
+		}
+	}
+	if !reflect.ValueOf(ref).IsZero() {
+		log.Printf("IDPool: GetID  Assigning key : %+v , id  %+v for ref %v", id, key, ref)
+		// refSet := ip._refs[id]
+		if reflect.ValueOf(ip._refs[id]).IsZero() {
+			ip._refs[id] = make([]interface{}, 0)
+		}
+		ip._refs[id] = append(ip._refs[id], ref)
+		return id, uint32(len(ip._refs[id]))
+	}
+	log.Printf("IDPool: GetID Assigning id %v for key %v and ref %v", id, key, ref)
+	return id, uint32(0)
+}
+
+func deleteRef(refSet []interface{}, ref interface{}) []interface{} {
+	// size := len(refSet)
+	var i uint32
+	for index, value := range refSet {
+		if value == ref {
+			i = uint32(index)
+			break
+		}
+	}
+	return append(refSet[:i], refSet[i+1:]...)
+}
+
+// ReleaseID get the reference id
+func (ip *IDPool) ReleaseID(key interface{}, ref interface{}) (uint32, uint32) {
+	// with self._lock:
+	log.Printf("IDPool:ReleaseID  Releasing id for key %v", key)
+	id := ip._idsInUse[key]
+	if reflect.ValueOf(ref).IsZero() {
+		log.Printf("No id to release for key %v", key)
+		return 0, 0
+	}
+	refSet := ip._refs[id]
+	if !reflect.ValueOf(refSet).IsZero() && !reflect.ValueOf(ref).IsZero() {
+		// Remove the specified reference from the id
+		refSet = deleteRef(refSet, ref)
+	}
+	if !reflect.ValueOf(refSet).IsZero() {
+		// No (remaining) references, release id
+		log.Printf("IDPool:ReleaseID Id %v has been released", id)
+		delete(ip._idsInUse, key)
+		if !reflect.ValueOf(refSet).IsZero() {
+			delete(ip._refs, id)
+		}
+		// Store released id for future reassignment
+		ip._idsForReuse[key] = id
+	} else {
+		log.Printf("IDPool:ReleaseID Keep id:%+v remaining references %+v", id, len(refSet))
+	}
+	if !reflect.ValueOf(ref).IsZero() {
```

Comment:
!= nil

Is this condition needed at all?

Reply:
Done